### ISSUE TITLE
distributed loss for multiple GPUs

### ISSUE THREAD

**Issue by x-zb** [luyug/GradCache#25]
Hi Luyu, thank you for your nice work.

I have a question on Distributed Contrastive Loss:
https://github.com/luyug/GradCache/blob/33695437d104e50a961cd9beba18b55c85a6537a/src/grad_cache/loss.py#L30-L34

It seems both `x` and `y` are gathered to all the GPUs. Then what is the meaning of distributed data parallel? For dense passage retrieval, we usually adopt the in-batch negatives technique, where only the passages (`y`) are gathered, not the queries (`x`). Is my understanding correct?

Besides, when I use multiple GPUs with Gradient Caching for training on Natural Questions, I got lower retrieval results (77.7 top-20 acc) than those in your paper and the original DPR paper (78.4 top-20 acc). Have you tried multiple GPUs in your case? Would appreciate it if you have any emiprical results.

**Comment by luyug** [comment]
The current implementation is a specific distributed compute design. In Pytorch, distributed operations are not differentiable so it is at the moment necessary to handcraft all communication for computing the gradients of the representations. The implementation in this repo materialize the full forward pass on all devices while keeping the backward sharded. This particular implementation avoids writing extra autograd function for communication in backward.

**Comment by x-zb** [comment]
I see. So in 
https://github.com/luyug/GradCache/blob/33695437d104e50a961cd9beba18b55c85a6537a/src/grad_cache/grad_cache.py#L193-L211
you only use `loss.backward()` on each GPU and do not need to communicate the gradients, right?
But will gathering `x` to all the GPUs consume more GPU memory than distrubting `x` to seperate GPUs? Is the memory difference significant?

**Comment by x-zb** [comment]
I see. So you mean when we use DDP and `loss.backward()`, `dist.all_gather()` cannot backpropogate gradients from the gathered tensors to the model parameters on other GPUs, right?

**Comment by luyug** [comment]
Effectively. The all_gather should give you leaf tensors, which won't get grads back-prop during the backward call.

**Comment by x-zb** [comment]
I see. Thanks for your explanations.
